src/apps/app_bulk_upload.erl

```diff
@@ -0,0 +1,212 @@
+-module(app_bulk_upload).
+-export([upload_file/2, download/2]).
+
+-include("ar.hrl").
+-include_lib("eunit/include/eunit.hrl").
+
+%%% Starts a server that accepts possibly huge files, splits them into small chunks, and
+%%% submits them to Arweave as separate transactions. Transactions are submitted via
+%%% the queue module.
+
+-define(CHUNK_SIZE, (1024 * 1024)).
+-define(BLOB_HASH_ALGO, sha256).
+
+%% @doc Starts a queue server, splits the given file into chunks, wraps the chunks as 
+%% Arweave transactions, and submits them to the queue. Returns the queue PID.
+upload_file(Wallet, Filename) ->
+	{ok, Filecontents} = file:read_file(Filename),
+	upload_blob(Wallet, Filecontents).
+
+upload_blob(Wallet, Blob) when is_binary(Blob) ->
+	upload_blob(whereis(http_entrypoint_node), Wallet, Blob).
+
+upload_blob(Node, Wallet, Blob) ->
+	Queue = app_queue:start(Node, Wallet),
+	upload_chunks(Queue, Blob),
+	Queue.
+
+%% @doc Takes a binary blob and processes it chunk by chunk. Each chunk is converted into
+%% a transaction and put into the queue. Chunk size is 1MB.
+upload_chunks(Queue, Blob) ->
+	upload_chunks(Queue, Blob, true).
+
+upload_chunks(Queue, Blob, IsFirstChunk) ->
+	case byte_size(Blob) =< ?CHUNK_SIZE of
+		true ->
+			app_queue:add(Queue, chunk_to_tx(Blob, IsFirstChunk));
+		false ->
+			<< Chunk:?CHUNK_SIZE/binary, Rest/binary >> = Blob,
+			app_queue:add(Queue, chunk_to_tx(Chunk, IsFirstChunk)),
+			upload_chunks(Queue, Rest, false)
+	end.
+
+%% @doc Converts the given binary chunk into a transaction.
+%% The first chunk is tagged as such so that we know where to stop when the blob is downloaded.
+chunk_to_tx(Chunk, IsFirstChunk) ->
+	#tx {
+		tags = chunk_tags(IsFirstChunk),
+		data = Chunk
+	}.
+
+chunk_tags(false) ->
+	[
+		{<< "app_name" >>, << "BulkUpload" >>}
+	];
+chunk_tags(true) ->
+	chunk_tags(false) ++ [{<< "first_chunk" >>, << "true" >>}].
+
+%% @doc Searches the local storage for the chunks of the blob identified by the given
+%% Base 64 encoded transaction ID. If the blob is reconstructed successfully,
+%% writes it to the specified destination.
+download(TXID, Filename) ->
+	{ok, Blob} = download(TXID),
+	file:write_file(Filename, Blob, [write]).
+
+download(TXID) ->
+	case download_chunks(TXID, []) of
+		Chunks when is_list(Chunks) ->
+			{ok, Chunks};
+		Err ->
+			{error, Err}
+	end.
+
+download_chunks(TXID, Chunks) ->
+	case ar_storage:read_tx(ar_util:decode(TXID)) of
+		unavailable ->
+			tx_not_found;
+		TX ->
+			case lists:keyfind(<< "app_name" >>, 1, TX#tx.tags) of
+				false ->
+					invalid_tx;
+				_ ->
+					case lists:keyfind(<< "first_chunk" >>, 1, TX#tx.tags) of
```

Comment:
See my comment for the parent case.

Btw, we don't like nested cases. How about casing on a tuple with both instead? Like: `{AppName, FirstChunk}`

Reply:
`"first_chunk"`'s value is not important so I do not check for it here. And yes, I will get rid of the nested case.